### DISCUSSION TITLE
setVal usage outside header file

### DISCUSSION THREAD

**Discussion by MauricioNO** [HomeSpan/HomeSpan#189] [Q&A] [answered]
Hello there! First I'd very much like to thank for the great work on HomeSpan, this is so well put together and documented!

I'll try to keep things simple to explain the situation: I have a switch that is defined in a DEV_Lock.h file 

    struct DEV_Lock : Service::Switch {              // Solenoid Lock    
    int lockPin;                                       // pin number defined for this LED
    int timeToLock;                                    // Time in ms holding the lock powered (open)
    
    SpanCharacteristic *power;                        // reference to the On Characteristic
    
    DEV_Lock(int lockPin, int timeToLock) : Service::Switch(){       // constructor() method
  
      power=new Characteristic::On(false);                 
          
      this->lockPin=lockPin;                            
      this->timeToLock=timeToLock;                          
      pinMode(lockPin,OUTPUT);                         
      
      Serial.print("Configuring Lock: Pin=");   // initialization message
      Serial.print(lockPin);
      Serial.print("  Time Held Open=");                      
      Serial.print(timeToLock);
      Serial.print("\n");
  
    } // end constructor`

This is how this switch is created in the main .ino file, inside void setup:

```
new SpanAccessory();                                                          
      (new DEV_Identify("Box","BoxM","01-20-L-RGB","20 box","1.0",0))->setPrimary();
      new DEV_Lock(LOCK_PIN,3000);
```

The thing is that this switch, in addition to be controlled via HomeKit, can also be triggered by an external action, such as an RFID tag. In my understanding, I'm supposed to handle this in the main loop() function, after proper tag validation, use the setVal function, right? So that the action on it can be handled directly by the specific boolean update() function defined in the DEV_Lock.h file.

The problem is that I can't (or better yet, don't know) how to make this call to the setVal from within the .ino file.
I tried a few combinations using dots and arrows (give a name to the Dev_Lock, as shown below) and try to call it:

```
new SpanAccessory();                                       
   (new DEV_Identify("Box","BoxM","01-20-L-RGB","20 box","1.0",0))->setPrimary();
   lockSol = new DEV_Lock(LOCK_PIN,3000);
```

-------
`lockSol->power->setVal(true)`

The problem is that I get an error message that 'lockSol'was not declared in this scope.
Any tips for me to try?

Thank you!

**Comment by MauricioNO** [comment]
Okay so I did some digging and research, and this is what I came to so far:

Declared before setup the following:
`DEV_Lock* lockSol;  `

Then inside setup have as I had in the snippet in the first post:
```
new SpanAccessory();                                       
   (new DEV_Identify("Box","BoxM","01-20-L-RGB","20 box","1.0",0))->setPrimary();
   lockSol = new DEV_Lock(LOCK_PIN,3000);
```

Then in the main loop function I'm able to call
`lockSol->power->setVal(true);`

Now I don't know yet if this will work as I expect, as I can't try it out now, maybe I'll get a chance tomorrow.
Since I'm no pointer expert, I'd appreciate if anybody understands what as tried out and can confirm it's a good approach :P 

Thanks!

**Comment by HomeSpan** [comment]
There are two separate `loop()` functions.  One is the main loop() function associated with the main sketch.  The other is a loop() method associated with each Service.  Every Service has a default loop() method that does nothing.  However, you can override this by creating a loop() method as part of your Service definition, in the same way that you create an update() method.  All the Characteristics in the Service will be available in the loop() method, including setVal().  See Example 12 for a detailed illustration of how the Service loop() method works.

**Reply by MauricioNO** [reply]
Yep I got that, this is my full DEV_Lock.h for reference:

```
struct DEV_Lock : Service::Switch {              // Solenoid Lock

  int lockPin;                                       // pin number defined for this LED
  int timeToLock;                                    // Time in ms holding the lock powered (open)
  
  SpanCharacteristic *power;                        // reference to the On Characteristic
  
  DEV_Lock(int lockPin, int timeToLock) : Service::Switch(){       // constructor() method

    power=new Characteristic::On(false);                 
        
    this->lockPin=lockPin;                            
    this->timeToLock=timeToLock;                          
    pinMode(lockPin,OUTPUT);                         
    
    Serial.print("Configuring Lock: Pin=");   // initialization message
    Serial.print(lockPin);
    Serial.print("  Time Held Open=");                      
    Serial.print(timeToLock);
    Serial.print("\n");

  } // end constructor

  boolean update(){                              // update() method

    if(power->getNewVal()){                       // check to ensure HomeKit is requesting we "turn on" this device (else ignore)

      LOG1("Activating the Solenoid lock on pin=");
      LOG1(lockPin);
      LOG1("\n");

      digitalWrite(lockPin,HIGH);                    // turn pin on - opens the lock

      
    } // if newVal=true
    
    return(true);                               // return true
  
  } // update


  void loop(){

    if(power->getVal() && power->timeVal()>timeToLock){   // check that power is true, and that time since last modification is greater than 3 seconds 
      LOG1("Releasing the lock\n");     // log message  
      power->setVal(false);                         // set power to false
      digitalWrite(lockPin,LOW);
    }      
    
  } // loop
  
};
```

The thing is that since not only HomeKit is able to trigger the switch, I don't see how updating the Service associated loop would help me out here.
My goal is to be able to call setVal from the main sketch loop function, as say the push of a button should give me the same result as making the call through HomeKit

**Comment by HomeSpan** [comment, marked as answer]
I just saw your update.  You can use the method you outlined to create a global lockSol variable, but placing all the logic inside the loop() method for the Service will be more robust and scalable (you can create multiple Locks, each with their own RFID).

**Comment by HomeSpan** [comment]
The RFID logic should be in the same loop function as well.  Within that loop you would additional check for an RFID scan.  If found, you trigger the lock and call setVal on power.  The nice thing is that your existing logic will handle the timeout and reset automatically, just as if you triggered the lock via HomeKit.

**Reply by MauricioNO** [reply]
Oh I think I see your point, basically put all the RFID code (pasted below for reference) inside the DEV_Lock.h loop. Yep I believe it should work given the loop function for the service gets called just the same, it could identify a new tag coming close from inside there.

I'll give that I try as well tomorrow and come back to confirm, thank you very much!

 ```
// RFID specific section
  if (rfid.PICC_IsNewCardPresent()) { // new tag is available
    if (rfid.PICC_ReadCardSerial()) { // NUID has been readed
      MFRC522::PICC_Type piccType = rfid.PICC_GetType(rfid.uid.sak);

      if (rfid.uid.uidByte[0] == keyTagUID[0] &&
          rfid.uid.uidByte[1] == keyTagUID[1] &&
          rfid.uid.uidByte[2] == keyTagUID[2] &&
          rfid.uid.uidByte[3] == keyTagUID[3] ) {
        Serial.println("Access is granted");

        digitalWrite(LOCK_PIN, HIGH);
   }
```

**Comment by HomeSpan** [comment]
Yes, you should be able to insert that directly into the Service loop() function - just add a setVal() and you should be set.  Let me know how it goes.

**Reply by MauricioNO** [reply]
It worked! Thank you!
